### PR TITLE
fix(azure_sentinel): split batches under the 1MB ingestion cap

### LINKED ISSUE #26450: [Bug]: Azure Sentinel logging fails due to Azure limits
### Check for existing issues

- [x] I have searched the existing issues and checked that my issue is not a duplicate.

### What happened?

Azure Data Collection Rule HTTP endpoint has a max limit of 1MB (raw or uncompressed), requests larger than this fail. Additionally, and field that is over 256KB is truncated, so there is no reason to try and send data larger than this

### Steps to Reproduce

1. Enable Azure sentinel logging
2. Create request with messages >1MB
3. It will not be logged in Azure, error will be in the LiteLLM logs

### Relevant log output

```shell
litellm-5f7c87bccb-dsdg2     response = await self.async_httpx_client.post(                                                                                
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                
         url=self.api_endpoint, data=body.encode("utf-8"), headers=headers                                                         
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                         
     )                                                                                                                             
     ^                                                                                                                             
   File "/usr/lib/python3.13/site-packages/litellm/litellm_core_utils/logging_utils.py", line 190, in async_wrapper                
     result = await func(*args, **kwargs)                                                                                          
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                          
   File "/usr/lib/python3.13/site-packages/litellm/llms/custom_httpx/http_handler.py", line 464, in post                           
     raise e                                                                                                                       
   File "/usr/lib/python3.13/site-packages/litellm/llms/custom_httpx/http_handler.py", line 420, in post                           
     response.raise_for_status()                                                                                                   
     ~~~~~~~~~~~~~~~~~~~~~~~~~^^                                                                                                   
   File "/usr/lib/python3.13/site-packages/httpx/_models.py", line 829, in raise_for_status                                        
     raise HTTPStatusError(message, request=request, response=self)                                                                
 httpx.HTTPStatusError: Client error '413 Request Entity Too Large' for url 'https://(REDACTED).ingest.monitor.azure.com/dataCollectionRules/(REDACTED)?api-version=2023-01-01'
 For more information check: https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/413
```

### What part of LiteLLM is this about?

Proxy

### What LiteLLM version are you on ?

1.83.13-nightly

### Twitter / LinkedIn details

_No response_

### PR DESCRIPTION
**Pull request by yucheng-berri**
## TLDR

Problem this solves:

- Azure Sentinel batches over 1MB get a 413 and are dropped whole
- Any send failure also dropped the whole queue, no retry
- Records added during an in-flight send were cleared unsent

How it solves it:

- Split batches under the 1MB ingestion cap before sending
- On 413, halve and retry, drop only a lone oversize record
- Keep undelivered records queued for the next flush, retried once per flush interval
- Take the flush lock for threshold sends too, so overlapping flushes stop double-sending records
- On a cancelled send, requeue only the pieces the destination never accepted
- Share the split logic with Datadog instead of copying it

## User Flow

Before: a proxy admin who ships request logs to Microsoft Sentinel sees a burst of large prompts vanish from Log Analytics, along with the small requests sent in the same few seconds

1. They send POST https://litellm-domain/v1/chat/completions six times at once, each with a 240KB prompt, plus one small request a few seconds earlier, and every call returns 200
2. They open the Log Analytics workspace and run `LiteLLM_CL | where TimeGenerated > ago(10m)`: none of the seven requests is there
3. The proxy log shows a single line, `Azure Sentinel Error sending batch API - Client error '413 Request Entity Too Large'`, and nothing is retried
4. The same happens for a short ingestion outage: one `503 Service Unavailable` line, and every record queued at that moment is gone for good

After: the same requests all show up in Log Analytics, and a short outage delays them instead of losing them

1. They send the same POST https://litellm-domain/v1/chat/completions six times at once with 240KB prompts, plus the one small request, and every call returns 200
2. `LiteLLM_CL | where TimeGenerated > ago(10m)` lists all seven requests
3. The proxy log shows no 413. A single request whose own record is over 1MB logs `Azure Sentinel API Error - Payload too large for a single record` and only that one record is skipped
4. During an ingestion outage the proxy logs one `503 Service Unavailable` line per flush, and once the endpoint is back the queued records land in Log Analytics on the next flush

## Relevant issues

Fixes #26450

## Linear ticket

Resolves LIT-5899

## Pre-Submission checklist

**Please complete all items before asking a LiteLLM maintainer to review your PR**

- [x] I have added meaningful tests
- [x] The handful of test files covering my change pass locally, e.g. `uv run pytest tests/test_litellm/<your_test_file>.py -v`. Leave the suites (`make test-unit-*`, `make test-unit`) to CI: it finishes in ~15 minutes where a laptop takes an hour or more
- [x] My PR passes all required CI/CD checks (e.g., lint, schema.d.ts sync check, etc.)
- [x] My PR's scope is as isolated as possible; it only solves 1 specific problem
- [x] I have received a Greptile **Confidence Score of at least 4/5** before requesting a maintainer review (Greptile reviews automatically once the PR is opened; only comment `@greptileai` to re-request a review after pushing changes)

## Screenshots / Proof of Fix

Datadog UI evidence: [public dashboard](https://p.us5.datadoghq.com/sb/9b1eff10-830b-11f0-b7cb-de02bd9ba50c-8bf13403bc3b8954210adbc5b61f68ac) and [dashboard screenshot](https://raw.githubusercontent.com/yucheng-berri/opentelemetry-python-contrib/assets-pr39880/lit5899/datadog-delivery-parity.png). The dashboard reads six real Datadog Logs events from both the base proxy and fixed head after the intake returned 403 for 30 seconds, then recovered

Azure portal evidence from the dedicated Log Analytics workspace:

Azure reactive 413 readback

Reactive live-413 readback: the forwarder returned 413 above 300KB and relayed smaller requests to the real Azure endpoint; the base delivered 1 of 6 records while the fixed head delivered all 6. [Open the full image](https://raw.githubusercontent.com/yucheng-berri/opentelemetry-python-contrib/assets-pr39880/lit5899/azure-reactive-413.png)

Azure proactive split readback

Real Azure proactive-split readback: the workspace contains every final-head large-batch and small-sibling record; the base's burst result varies with flush timing and dropped mixed oversize batches in the recorded run. [Open the full image](https://raw.githubusercontent.com/yucheng-berri/opentelemetry-python-contrib/assets-pr39880/lit5899/azure-proactive-split.png)

Azure audit delivery readback

Audit stream parity readback: both base and head delivered the small audit record, confirming the audit callback path remains intact. [Open the full image](https://raw.githubusercontent.com/yucheng-berri/opentelemetry-python-contrib/assets-pr39880/lit5899/azure-audit-delivery.png)

Azure portal UI steps: open `https://portal.azure.com`, select Log Analytics workspaces, open `lit5899-ws` in resource group `litellm-lit5899-rg`, choose Logs, paste the `LiteLLM_CL` query below, set the time range to Last hour, and run it. The query shows the final-head `LIT5899_R6_S1_HEAD`, `LIT5899_R6_S3_HEAD`, and `LIT5899_R6_S4SMALL_HEAD` rows. Run the `LiteLLMAudit_CL` query below for the audit marker

Shared setup, identical for both runs. Real proxy on a fresh Postgres, real OpenAI, real Azure Monitor Logs Ingestion endpoint (a dedicated Log Analytics workspace with a `Custom-LiteLLM` and a `Custom-LiteLLM-Audit` stream), one worker

```yaml
model_list:
  - model_name: gpt-4o-mini
    litellm_params:
      model: openai/gpt-4o-mini
      api_key: os.environ/OPENAI_API_KEY

litellm_settings:
  callbacks: ["azure_sentinel"]
  audit_log_callbacks: ["azure_sentinel"]
  store_audit_logs: true

general_settings:
  master_key: sk-lit5899
```

```bash
export AZURE_SENTINEL_ENDPOINT=... AZURE_SENTINEL_DCR_ID=... AZURE_SENTINEL_TENANT_ID=... AZURE_SENTINEL_CLIENT_ID=... AZURE_SENTINEL_CLIENT_SECRET=...
export DATABASE_URL=postgresql://litellm:litellm@127.0.0.1:15899/lit5899
python litellm/proxy/proxy_cli.py --config config.yaml --port 4000

# payloads: a 216 byte prompt, a 240KB prompt, a 1.1MB prompt, each carrying a marker like LIT5899_R2_S3_<BASE|HEAD>
python -c 'import json,sys; print(json.dumps({"model":"gpt-4o-mini","max_tokens":5,"messages":[{"role":"user","content":sys.argv[1]+" "+"x"*int(sys.argv[2])}]}))' LIT5899_R2_S3_HEAD 240000 > s3.json

# readback, run a minute after the last request
WS=<workspace id>
az monitor log-analytics query -w $WS -o table --analytics-query "LiteLLM_CL | where TimeGenerated > ago(1h) | extend marker = extract('LIT5899_R2_[A-Z0-9]+_(?:BASE|HEAD)', 0, tostring(messages)) | where isnotempty(marker) | summarize records = count(), firstSeen = min(TimeGenerated) by marker | order by marker asc"
az monitor log-analytics query -w $WS -o table --analytics-query "LiteLLMAudit_CL | where TimeGenerated > ago(1h) | extend marker = extract('LIT5899_R2_S5_(?:base|head)', 0, tostring(updated_values)) | where isnotempty(marker) | summarize records = count(), firstSeen = min(TimeGenerated) by marker, action, table_name | order by marker asc"
```

#### One small request followed by six concurrent 240KB requests

1. `curl -sS -o /dev/null -w 'status=%{http_code}\n' http://127.0.0.1:4000/v1/chat/completions -H 'Authorization: Bearer sk-lit5899' -H 'Content-Type: application/json' -d @s1.json`

   ```
   status=200
   ```

2. `seq 1 6 | xargs -P 6 -I{} curl -sS -o /dev/null -w 'status=%{http_code}\n' http://127.0.0.1:4000/v1/chat/completions -H 'Authorization: Bearer sk-lit5899' -H 'Content-Type: application/json' -d @s3.json | sort | uniq -c`

   ```
      6 status=200
   ```

3. Proxy log at the next flush

   ```
   01:17:35 - LiteLLM:ERROR: azure_sentinel.py:372 - Azure Sentinel Error sending batch API - Client error '413 Request Entity Too Large' for url '<url>
   ```

4. `LiteLLM_CL` readback: no rows for `LIT5899_R2_S1_BASE` or `LIT5899_R2_S3_BASE`. All seven records were dropped, including the small one that happened to share the batch

#### One 1.1MB request plus three small ones

1. `curl ... -d @s4big.json & seq 1 3 | xargs -P 3 -I{} curl ... -d @s4small.json; wait`

   ```
      1 big status=400
      3 small status=200
   ```

   The 400 is OpenAI rejecting the 1.1MB prompt, which the proxy still logs as a failure record

2. Proxy log at the next flush

   ```
   01:17:40 - LiteLLM:ERROR: azure_sentinel.py:372 - Azure Sentinel Error sending batch API - Client error '413 Request Entity Too Large' for url '<url>
   ```

3. `LiteLLM_CL` readback: no rows for `LIT5899_R2_S4SMALL_BASE`. The three small records were dropped along with the one oversize record

#### Team creation audit record

1. `curl -sS -o /dev/null -w 'status=%{http_code}\n' http://127.0.0.1:4000/team/new -H 'Authorization: Bearer sk-lit5899' -H 'Content-Type: application/json' -d '{"team_alias":"LIT5899_R2_S5_base"}'`

   ```
   status=200
   ```

2. `LiteLLMAudit_CL` readback

   ```
   TableName      Action    FirstSeen                     Marker              Records    Table_name
   PrimaryResult  created   2026-09-05T08:19:31.2674302Z  LIT5899_R2_S5_base  1          LiteLLM_TeamTable
   ```

   Small audit batches were never affected, this case is the parity check

#### Ingestion endpoint answers 503 for 25 seconds, then recovers

1. Point `AZURE_SENTINEL_ENDPOINT` at a local forwarder that answers 503 while `/tmp/fault.flag` exists and otherwise relays the request unchanged to the real endpoint. `touch /tmp/fault.flag`, then send three chat completions with marker `LIT5899_REC_base` and one `POST /team/new` with alias `LIT5899_REC_S5_base`

   ```
   base chat 1 status=200
   base chat 2 status=200
   base chat 3 status=200
   base team/new status=200
   ```

2. Forwarder log after 22 seconds, then `rm /tmp/fault.flag` and 15 more seconds

   ```
      1 FAULT 503 bytes=35800 path=.../streams/Custom-LiteLLM?api-version=2023-01-01
      1 FAULT 503 bytes=614 path=.../streams/Custom-LiteLLM-Audit?api-version=2023-01-01
   ```

   One attempt per stream, nothing after the fault is lifted

3. Proxy log

   ```
   01:14:36 - LiteLLM:ERROR: azure_sentinel.py:372 - Azure Sentinel Error sending batch API - Server error '503 Service Unavailable' for url '<url>
   01:14:36 - LiteLLM:ERROR: azure_sentinel.py:372 - Azure Sentinel Error sending batch API - Server error '503 Service Unavailable' for url '<url>
   ```

4. Readback with the `LIT5899_REC_` markers: no rows in `LiteLLM_CL`, no rows in `LiteLLMAudit_CL`

#### One small request followed by six concurrent 240KB requests

1. `curl -sS -o /dev/null -w 'status=%{http_code}\n' http://127.0.0.1:4000/v1/chat/completions -H 'Authorization: Bearer sk-lit5899' -H 'Content-Type: application/json' -d @s1.json`

   ```
   status=200
   ```

2. `seq 1 6 | xargs -P 6 -I{} curl -sS -o /dev/null -w 'status=%{http_code}\n' http://127.0.0.1:4000/v1/chat/completions -H 'Authorization: Bearer sk-lit5899' -H 'Content-Type: application/json' -d @s3.json | sort | uniq -c`

   ```
      6 status=200
   ```

3. Proxy log at the next flush: no error line

4. `LiteLLM_CL` readback

   ```
   TableName      FirstSeen                     Marker                   Records
   PrimaryResult  2026-09-05T08:18:10.9170946Z  LIT5899_R2_S1_HEAD       1
   PrimaryResult  2026-09-05T08:18:21.5206178Z  LIT5899_R2_S3_HEAD       6
   ```

#### One 1.1MB request plus three small ones

1. `curl ... -d @s4big.json & seq 1 3 | xargs -P 3 -I{} curl ... -d @s4small.json; wait`

   ```
      1 big status=400
      3 small status=200
   ```

2. Proxy log at the next flush

   ```
   01:18:37 - LiteLLM:ERROR: batch_utils.py:22 - Azure Sentinel API Error - Payload too large for a single record
   ```

3. `LiteLLM_CL` readback

   ```
   TableName      FirstSeen                     Marker                   Records
   PrimaryResult  2026-09-05T08:18:36.8969782Z  LIT5899_R2_S4SMALL_HEAD  3
   ```

   The three small records land. Only the single record that is over 1MB on its own is skipped, and the log names it

#### Team creation audit record

1. `curl -sS -o /dev/null -w 'status=%{http_code}\n' http://127.0.0.1:4000/team/new -H 'Authorization: Bearer sk-lit5899' -H 'Content-Type: application/json' -d '{"team_alias":"LIT5899_R2_S5_head"}'`

   ```
   status=200
   ```

2. `LiteLLMAudit_CL` readback

   ```
   TableName      Action    FirstSeen                     Marker              Records    Table_name
   PrimaryResult  created   2026-09-05T08:19:32.441537Z   LIT5899_R2_S5_head  1          LiteLLM_TeamTable
   ```

#### Ingestion endpoint answers 503 for 25 seconds, then recovers

1. Same forwarder, `touch /tmp/fault.flag`, three chat completions with marker `LIT5899_REC_head` and one `POST /team/new` with alias `LIT5899_REC_S5_head`

   ```
   head chat 1 status=200
   head chat 2 status=200
   head chat 3 status=200
   head team/new status=200
   ```

2. Forwarder log after 22 seconds, then `rm /tmp/fault.flag` and 15 more seconds

   ```
      5 FAULT 503 bytes=35799 path=.../streams/Custom-LiteLLM?api-version=2023-01-01
      5 FAULT 503 bytes=614 path=.../streams/Custom-LiteLLM-Audit?api-version=2023-01-01
      1 FORWARD status=204 bytes=35799 path=.../streams/Custom-LiteLLM?api-version=2023-01-01
      1 FORWARD status=204 bytes=614 path=.../streams/Custom-LiteLLM-Audit?api-version=2023-01-01
   ```

   One retry per flush interval while the fault is on, then both batches delivered with 204 as soon as it lifts

3. Proxy log

   ```
   01:14:36 - LiteLLM:ERROR: batch_utils.py:55 - Azure Sentinel Error sending batch API - Server error '503 Service Unavailable' for url '<url>
   01:14:41 - LiteLLM:ERROR: batch_utils.py:55 - Azure Sentinel Error sending batch API - Server error '503 Service Unavailable' for url '<url>
   01:14:46 - LiteLLM:ERROR: batch_utils.py:55 - Azure Sentinel Error sending batch API - Server error '503 Service Unavailable' for url '<url>
   01:14:51 - LiteLLM:ERROR: batch_utils.py:55 - Azure Sentinel Error sending batch API - Server error '503 Service Unavailable' for url '<url>
   01:14:56 - LiteLLM:ERROR: batch_utils.py:55 - Azure Sentinel Error sending batch API - Server error '503 Service Unavailable' for url '<url>
   ```

   (one pair per flush, the audit stream's twin lines omitted for brevity)

4. Readback with the `LIT5899_REC_` markers

   ```
   TableName      FirstSeen                     Marker            Records
   PrimaryResult  2026-09-05T08:15:01.8701628Z  LIT5899_REC_head  3

   TableName      FirstSeen                     Marker               Records
   PrimaryResult  2026-09-05T08:15:02.0582746Z  LIT5899_REC_S5_head  1
   ```

### Retry cadence and concurrent threshold sends with `DEFAULT_BATCH_SIZE=4` (396f0ee341)

Same forwarder, with `DEFAULT_BATCH_SIZE=4` exported on all proxies. The new leg sent concurrent requests while the endpoint returned 503, then read back the real workspace

1. Twelve concurrent requests per side, one worker, 22 seconds under fault, then 15 seconds after recovery

   ```
   Forwarder 503 attempts: base 9, prev 9, head 5
   Log Analytics records:  base 0, prev 12, head 12
   ```

2. Forty-eight concurrent requests per side, four workers, 22 seconds under fault, then 15 seconds after recovery

   ```
   Forwarder 503 attempts: base 32, prev 45, head 21
   Log Analytics records:  base 0, prev 48, head 48
   Workers started:       4 per proxy
   ```

   The previous head let concurrent threshold waiters resend the growing queue after the first failure. The new head rechecks the retry flag and threshold under the flush lock, so one failed threshold attempt is followed by timer retries only, and recovery delivers every record exactly once

3. The earlier sequential leg remains covered: base indexed 17 records for 9 requests under overlapping flushes, while the fixed head indexed exactly 9; during a 22-second outage, the fixed head retried once per 5-second flush interval and delivered all standard and audit records after recovery

### Reactive halving on a live 413 (4b78699ba7)

Azure's cap equals litellm's 1MB estimate, so the real destination never exercises the halve-on-413 branch. For this leg the forwarder answered 413 for any body over 300KB and relayed smaller bodies to the real ingestion endpoint. Six concurrent 240KB requests per side, `DEFAULT_BATCH_SIZE=8`

```
Forwarder, base:  413 bytes=1261033 (five records, dropped), 204 bytes=252237 (the straggler that flushed alone)
Forwarder, head:  413 bytes=756608, 413 bytes=756624, 413 bytes=504411, 413 bytes=504430, then six 204 forwards of ~252KB each

Log Analytics:    LIT5899_R7D_S3_BASE 1 record, LIT5899_R7D_S3_HEAD 6 records
```

The head split the six records into two 756KB halves, each 413 was halved again, and every single-record body was accepted; no error line was logged. The base dropped its five-record batch on the first 413

## Type

🐛 Bug Fix

### Medium

- A single record over 1MB on its own is still dropped, with an error line naming it
- Records held across a long outage are capped at `max_queue_size` (default 50000), oldest dropped first

### Low

- While records are waiting for a retry, hitting `batch_size` no longer triggers a send; the periodic flush is the only retry path, so a failed backlog is attempted once per `flush_interval` (default 5s) instead of once per new event
- Other `CustomBatchLogger` subclasses still decide their batch-size sends outside the flush lock, and this PR only changes Azure Sentinel, so that stays for a follow-up
- A record that cannot be serialized (a mixed-type set in the payload, for example) is dropped on its own with an error line; before this PR it raised out of the flush and stopped the periodic flush task
- Datadog now logs `Datadog delivered N records` and `Datadog API error: status_code=`
  - the previous texts were `Datadog: delivered N events` and `Datadog: unexpected response status_code=`
  - anything alerting on the old literal text needs the new pattern
- Permanent 4xx responses other than 413, including OAuth token failures, are dropped instead of retried forever for Sentinel; Datadog keeps its existing requeue behavior for every non-413 HTTP failure, including permanent 4xx responses
- The shared helper defaults to requeueing on a non-413 failure, so a future integration that adopts it keeps its records unless it asks for dropping; Sentinel passes `undelivered_after_http_error` explicitly
- Sentinel recovery requests are capped at `batch_size` records, and cancellation requeues only the unsent suffix

### Final exhaustive live matrix (4b78699ba7)

The final head was re-driven base-versus-head across every remaining Azure Sentinel and Datadog branch against real Azure Monitor and Datadog destinations. HTTP faults were injected only at destination or OAuth boundaries

- Sentinel: permanent 401 and 403 drops; transient 408, 429, 503, malformed-token, whole-send cancellation, and mid-split cancellation recovery; independent standard and audit queues; zero-sized retry cap; unserializable-record isolation; proactive and reactive 413 splitting; fresh clean-marker real-Azure delivery
- Datadog: terminal 413; non-413 429 and 503 compatibility; reactive and proactive splitting; whole-send and mid-split cancellation; unserializable-record isolation; real Logs API readback
- Readback: head recovered four distinct records from each transient OAuth leg, all four whole-cancel Sentinel records, all four mid-split Sentinel records once, four standard plus one audit record after a shared 503, all 1 + 6 + 3 clean-marker Azure records, six reactive-split Datadog records per side, all four and six Datadog cancellation records, and exactly three good poison-sibling events
- Base controls reproduced the expected loss or poison wedge; Datadog response and split ordering stayed compatible. No changed live path remains untested and no regression or backward-incompatible risk was found

## Final Attestation

- [x] The tests check the right things, including the edge cases, and regressions in the respective real-world customer use-cases are not possible after this PR

---

> [!NOTE]
> **Medium Risk**
> Changes core logging delivery semantics for Azure Sentinel (retry vs drop on 4xx, queue behavior under failure) and refactors Datadog’s batch send path; misconfiguration could now drop permanent 4xx batches in Sentinel where Datadog still requeues.
> 
> **Overview**
> Fixes Azure Sentinel (and hardens Datadog) so proxy logs are not lost when batches exceed Azure Monitor’s ~1MB limit or when ingestion/OAuth fails transiently.
> 
> **Shared `batch_utils`:** New `send_batch_with_413_split` proactively splits oversized batches, halves on 413, drops only a single record that still 413s, isolates unserializable records, and on cancel requeues only the suffix that was never accepted. HTTP handling distinguishes retryable errors (5xx, 408, 429) from permanent 4xx via `undelivered_after_http_error` (Sentinel) vs Datadog’s existing “requeue all non-413” behavior.
> 
> **Azure Sentinel:** Sends through the shared helper with `AZURE_SENTINEL_MAX_PAYLOAD_SIZE_BYTES` (1MB) and per-request `batch_size` caps. Queues are **detached** before send and **merged back** with undelivered items (with `max_queue_size` trimming). **`logs_awaiting_retry` / `audit_logs_awaiting_retry`** block batch-size triggers while a backlog is pending so retries run on the periodic flush only. Threshold sends use **`flush_lock`** with rechecks to avoid duplicate sends and ordering bugs under concurrency.
> 
> **Datadog:** Replaces inline 413/split loop with the same helper; adds tests for serialization failures and mid-split cancellation.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 4b78699ba75c2c531127fe21dc4a29a14ba58cb5. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

ran /live-pr-risk and found no regressions/backward incompatible risks

### CODE REVIEW

File: litellm/integrations/azure_sentinel/azure_sentinel.py [hunk 1]
```diff
@@ -311,67 +315,82 @@ async def async_send_batch(self):
...
+        batch_to_send: Final = tuple(self.log_queue)
+        self.log_queue.clear()
+        undelivered: Final = await self._async_send_batch_to_api(
+            log_queue=batch_to_send,
             api_endpoint=self.api_endpoint,
             log_type="logs",
         )
+        if undelivered:
+            self.log_queue = list(undelivered) + self.log_queue
+            self._drop_oldest_over_max_queue_size(self.log_queue, "logs")
```
[the remark this answers is not in the document]
Comment (yucheng-berri) [reply litellm/integrations/azure_sentinel/azure_sentinel.py]: Fixed in ef94b7d3c1: threshold sends now go through flush_queue, so they take the same lock as the timer and cannot interleave

File: litellm/integrations/azure_sentinel/azure_sentinel.py [hunk 2]
```diff
@@ -311,67 +315,82 @@ async def async_send_batch(self):
         Raises:
             Raises a NON Blocking verbose_logger.exception if an error occurs
         """
-        await self._async_send_batch_to_api(
-            log_queue=self.log_queue,
+        batch_to_send: Final = tuple(self.log_queue)
+        self.log_queue.clear()
+        undelivered: Final = await self._async_send_batch_to_api(
```
[the remark this answers is not in the document]
Comment (yucheng-berri) [reply litellm/integrations/azure_sentinel/azure_sentinel.py]: Fixed in ef94b7d3c1: both queues are rebound now, and the requeue helper builds a new list instead of mutating one

File: litellm/integrations/azure_sentinel/azure_sentinel.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (yucheng-berri) [reply litellm/integrations/azure_sentinel/azure_sentinel.py]: Fixed in ef94b7d3c1: an awaiting-retry flag stands threshold sends down, so a failed backlog retries once per flush interval

File: litellm/integrations/azure_sentinel/azure_sentinel.py [hunk 3]
```diff
@@ -311,67 +315,82 @@ async def async_send_batch(self):
...
+            log_queue=batch_to_send,
             api_endpoint=self.audit_api_endpoint,
             log_type="audit logs",
         )
+        if undelivered:
+            self.audit_log_queue = list(undelivered) + self.audit_log_queue
+            self._drop_oldest_over_max_queue_size(self.audit_log_queue, "audit logs")
+
+    def _drop_oldest_over_max_queue_size(self, queue: list[_QueuedPayload], log_type: str) -> None:
+        overflow: Final = len(queue) - self.max_queue_size
```
[the remark this answers is not in the document]
Comment (yucheng-berri) [reply litellm/integrations/azure_sentinel/azure_sentinel.py]: max_queue_size is the CustomBatchLogger contract every batch logger shares, and operators can lower it. Repeated serialization is gone, retries wait for the timer

**Comment by yucheng-berri** [comment]
@greptileai please review the current head ef94b7d3

**Comment by yucheng-berri** [comment]
bugbot run

File: litellm/integrations/azure_sentinel/azure_sentinel.py [hunk 4]
```diff
@@ -245,8 +251,8 @@ async def async_log_success_event(self, kwargs, response_obj, start_time, end_ti
 
             self.log_queue.append(standard_logging_payload)
 
-            if len(self.log_queue) >= self.batch_size:
-                await self.async_send_batch()
+            if len(self.log_queue) >= self.batch_size and not self.logs_awaiting_retry:
+                await self.flush_queue()
```
[the remark this answers is not in the document]
Comment (yucheng-berri) [reply litellm/integrations/azure_sentinel/azure_sentinel.py]: Fixed in 396f0ee341: threshold paths recheck retry state under the lock, isolate standard and audit queues, and leave retries to the timer

**Comment by yucheng-berri** [comment]
@greptileai please review the current head 396f0ee341

**Comment by yucheng-berri** [comment]
bugbot run

**Comment by yucheng-berri** [comment]
@greptileai please review the current head add9ef4083

**Comment by yucheng-berri** [comment]
bugbot run

File: litellm/integrations/azure_sentinel/azure_sentinel.py [hunk 5]
```diff
@@ -298,80 +304,121 @@ async def async_log_audit_log_event(self, audit_log: StandardAuditLogPayload) ->
...
+        self.log_queue = []
+        try:
+            undelivered: Final = await self._async_send_batch_to_api(
+                log_queue=batch_to_send,
+                api_endpoint=self.api_endpoint,
+                log_type="logs",
+            )
+        except asyncio.CancelledError:
+            self.logs_awaiting_retry = True
+            self.log_queue = self._requeue(batch_to_send, self.log_queue, "logs")
```
[the remark this answers is not in the document]
Comment (yucheng-berri) [reply litellm/integrations/azure_sentinel/azure_sentinel.py:350]: Confirmed and fixed in 577ff5c67c: the split now raises a cancellation carrying only undelivered records, and a test covers the half delivered case

**Comment by yucheng-berri** [comment]
@greptileai please review the current head 577ff5c67c

**Comment by yucheng-berri** [comment]
bugbot run

**Comment by yucheng-berri** [comment]
@greptileai please review the current head 0b0e4701de

**Comment by yucheng-berri** [comment]
bugbot run

**Comment by yucheng-berri** [comment]
@greptileai please review the current head 16e2e80407

**Comment by yucheng-berri** [comment]
bugbot run

**Comment by yucheng-berri** [comment]
@greptileai please review the current head 154ad28fbf

**Comment by yucheng-berri** [comment]
bugbot run

File: litellm/integrations/batch_utils.py [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+        )
+        return tuple(batch)
+    verbose_logger.error(
+        "%s API error: status_code=%s is not retryable, dropped %s records - %s",
+        integration_name,
+        status_code,
+        len(batch),
+        detail,
+    )
+    return ()
```
[the remark this answers is not in the document]
Comment (yucheng-berri) [reply litellm/integrations/batch_utils.py:43]: Fixed in 154ad28fbf: Datadog now passes requeue_after_http_error, so every non-413 HTTP failure is retained as before

File: litellm/integrations/datadog/datadog.py [hunk 7]
```diff
@@ -413,53 +416,16 @@ async def _send_with_413_split(self, batch: list) -> list:
...
-    def _undelivered(chunk: list, pending: list[list]) -> list:
-        return chunk + [event for remaining in reversed(pending) for event in remaining]
+        undelivered: Final = await send_batch_with_413_split(
+            batch=batch,
+            send_batch=self.async_send_compressed_data,
+            exceeds_limits=self._exceeds_intake_limits,
+            success_status_codes=frozenset({202}),
+            integration_name="Datadog",
+            drop_error_message=DD_ERRORS.DATADOG_413_ERROR.value,
+            non_success_handler=requeue_after_http_error,
```
[the remark this answers is not in the document]
Comment (yucheng-berri) [reply litellm/integrations/datadog/datadog.py:426]: Live base and head 403 legs both requeued records and recovered them after fault removal; queue bounding predates this PR and remains a follow-up

**Comment by yucheng-berri** [comment]
@greptileai please review the current head 4b78699ba7, which makes requeue the shared helper default and has Sentinel opt into dropping explicitly

**Comment by yucheng-berri** [comment]
bugbot run

File: litellm/integrations/batch_utils.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (yucheng-berri) [reply litellm/integrations/batch_utils.py:43]: Done in 4b78699ba7: the helper now defaults to requeue_after_http_error and Sentinel passes undelivered_after_http_error explicitly, so no caller inherits dropping

**Review by tin-berri — APPROVED** [review]
